tools/socket_passing.py

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env python
+
+# This tool is a helper script that queries the admin address for all listener
+# addresses after envoy startup. The script can then be used to update an
+# exisiting json config file with updated listener addresses. This script is
+# currently called in the hot restart integration test to update listener
+# addresses bound to port 0 in the intial json config file. With the
+# -n or -no_port_change option, this script does not update an exisiting json
+# config file but checks that the listener addresses after envoy startup
+# match the listener addresses in the initial json config file.
+
+from collections import OrderedDict
+
+import argparse
+import httplib
+import json
+import os.path
+import sys
+import time
+
+ADMIN_FILE_TIMEOUT_SECS = 20
+
+def ReplaceListenerAddresses(original_json, admin_address, updated_json):
+  # Get original listener addresses
+  with open(original_json, 'r') as original_json_file:
+    # Import original config file in order to get a deterministic output. This
+    # allows us to diff the original config file and the updated config file
+    # output from this script to check for any changes.
+    parsed_json = json.load(original_json_file, object_pairs_hook=OrderedDict)
+  original_listeners = parsed_json['listeners']
+
+  sys.stdout.write('Admin address is ' + admin_address + '\n')
+  try:
+    admin_conn = httplib.HTTPConnection(admin_address)
+    admin_conn.request('GET', '/listeners')
+    admin_response = admin_conn.getresponse()
+    if not admin_response.status == 200:
+      return False
+    discovered_listeners = json.loads(admin_response.read())
+  except:
+    sys.stderr.write('Cannot connect to admin.\n')
```

Comment:
Generally it's scary to do a wildcard exception catch in Python and then not propagate further. In this case, it means exceptions will be hard to debug as we lose info. You could do `except e: \n sys.stderr.write('Cannot connect to admin: %s' % e)` or something like that. Or re-raise with

```
except e:
  sys.stdout.write(...)
  raise e
```

or create a custom exception class. Probably the first is fine for this test tool.